src/instructlab/sdg/generate_data.py

```diff
@@ -290,6 +311,8 @@ def generate_data(
         model_family,
         model_name,
         num_instructions_to_generate,
+        num_workers=num_cpus,
+        batch_size=batch_size,
```

Comment:
Hm, good point. This could get tricky to estimate. I think the primary elements of a heuristic would be expected memory overhead per batch element (likely low for IO-bound work), and expected concurrency limits on the client-side for IO-bound work. I'm not sure we would have any reasonable way of knowing these though. I have much less familiarity with the "real" workload, though. Is there some kind of fixed number that we think would make sense?

The other concern I have about enabling concurrency by default is the possibility for the sharding of the dataset to actually change the results. I _think_ most blocks will treat rows of the `Dataset` as independent entities and process them as such, but I could certainly imagine a case where the logic of a block does not treat the rows as independent. In this case, enabling concurrency might degrade the overall results if the individual block executions have less full context to work with.

Reply:
Our users certainly aren't going to be able to do a better job of judging these questions (whether sharding is safe, what batch size to us)

If we can't come up with a sane default, it should be in the pipeline config somehow

@shivchander @xukai92 @aakankshaduggal @npalaska we need your help making a call on an appropriate default batch size or whether it needs to be configured per-pipeline or per-block

Reply:
From @shivchander 

> we did some tests with the vllm endpoint and found that low batch size and increasing the number of workers worked best from a throughput perspective. So I’d recommend a batch size of 8 and setting num workers to num cpus.

Reply:
Further from @shivchander - it can be tricky to debug with this concurrency enabled, so would be nice to have an easy way to disable this for debugging (as a runtime parameter, not a pipeline config). It might make sense for `--num-cpus=1` to disable this?

Reply:
Yep - another reason would be for smaller datasets and small knowledge sources (say 1 page document), batching is an overkill and would probably end up taking more time due to I/O and regrouping. So I would incline towards `--num-cpus=1` as default behavior and triggering the batching when needed

Reply:
Ok, so it sounds like we want:

* `batch_size=8`
* `num_cpus=None` (<- The python default with `num_workers=None` is [here](https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor))
* If the user sets `num_cpus=1`, we fully disable using `ThreadPoolExecutor`
    * This one can get annoying since there is no out-of-the-box `SynchronousExecutor` that offers the same API. I wrote one in a different project that we can probably borrow ([here](https://github.com/IBM/oper8/blob/main/oper8/dag/runner.py#L294)). The alternative is to maintain separate code paths for the synchronous version which can be pretty error prone.

Reply:
> Yep - another reason would be for smaller datasets and small knowledge sources (say 1 page document), batching is an overkill and would probably end up taking more time due to I/O and regrouping. So I would incline towards `--num-cpus=1` as default behavior and triggering the batching when needed

My sense is that this will inevitably lead to users experiencing sub-par performance, complaining (or just walking way), and us telling them "oh, you need this go-faster flag"

I don't think we should be taking that route out of a concern that it might be a little slower with small datasets

Unless I'm misunderstanding something, I think we should enable the batching by default

Reply:
That's a fair point - im with you on making the distributed setting as a default 
```
batch_size = 8
num_cpus =  mp.cpu_count()
```

Reply:
Also, just a note: The combination of batch size and concurrency also depends on the underlying accelerator being used and how the accelerators can handle the batches. (The total number of batches the server can handle depends on the compute profile and the memory available on the accelerator). 

The general consensus is to have smaller batches per thread and maximize the number of threads to a point where we are not adding a lot of requests in the backend vLLM queue. As the requests start to queue up on the backend vLLM server the time to the first token will shoot up and some requests will result in timeout errors. However, having a small batch means only a few requests get penalized in case of an error.

For example, using a batch of 8 and num_workers 32 (256 total concurrent requests) on 4H100 yielded the most optimal performance for me.

Reply:
Thanks for the input! This makes sense and absolutely lines up with my assumptions about why a heuristic for this would be tough to nail down. Since we do have some strong ideas about the default environment where this would run, I think `batch_size = 8` and `num_cpus = None` (thereby using the python default of `min(32, os.cpu_count() + 4)`) is probably a safe place to start.